README.md
# Assertive for Promises

**Assertive as Promised** extends [Assertive][assertive] for asserting things
about standards-compliant promises.  It is 100% backward-compatible, so all of
the existing assertive documentation applies.

# DEPRECATED

As of `assertive-2.1.0`, this functionality has been rolled into
[assertive] itself.  The only functional change is that you can no longer
pass functions-which-return-promises as arguments to `resolves()` and
`rejects()` - you should replace things like:

```coffee
assert = require 'assertive-as-promised'
assert.rejects -> doSomething('blah')
```

with

```coffee
assert = require 'assertive'

# this version will explode your test if there is a synchronous throw
# during the invocation of doSomething() (which may be a good thing -
# you can test for that behavior separately with assert.throws())
assert.rejects(doSomething('blah')).then (err) -> ...

# this version will turn any synchronous throw into a rejection which will
# be available as err
Promise = require 'bluebird'
assert.rejects(Promise.try -> doSomething('blah')).then (err) ->
```

## How to Use

This is best used with something like [Mocha] (version >= 1.18.0) which
handles returned promises correctly.  All of assertive's assertions are
extended to accept promises as their argument to be tested and return a
promise which will be resolved or rejected.

For all existing assertive functions, you may simply replace the
argument with a promise for an equivalent argument.
`assert.equal('foo', funcThatReturnsAString())` becomes
`assert.equal('foo', funcThatReturnsAPromiseForAString())`.  Note that you may
get nicer and more consistent errors if you put any function calls that may
have a risk of throwing an exception synchronously inside a bluebird
`Promise.try()`, thusly:
`assert.equal('foo', Promise.try -> funcThatReturnsAPromiseForAString())`

Note for `throws()` and `notThrows()` that they accept a function (which may
throw a *synchronous* exception) or a promise for a function (which
may throw a *synchronous* exception).  The resolution status of the promise
itself is not being tested.  Since you're often more interested in the
resolution status of the promise, there are two new functions:
`rejects` and `resolves`:

`assert.rejects(-> funcThatReturnsAPromise(someArg))` takes as its
argument a promise OR a function that returns a promise.  The equivalent
counterpart to `notThrows` is called `resolves`.  `rejects` returns a promise
for the rejection error, and thus composes nicely with other assertions.

## Examples (using Mocha)

```coffee
{ runSync, runAsync }  = require './some-library'
assert  = require 'assertive-as-promised'
Promise = require 'bluebird'
# runAsync returns a promise

it 'runs synchronously', ->
  assert.deepEqual 'got proper hash', { a: 42 }, runSync('good')

it 'fails synchronously', ->
  assert.throws 'fails on bad', -> runSync('bad')

it 'runs asynchronously', ->
  assert.deepEqual 'got proper hash', { a: 42 }, runAsync('good')

it 'fails asynchronously', ->
  assert.rejects 'fails on bad', -> runAsync('bad')

fn = -> Promise.try -> throw 'kaboom'
it 'fails asynchronously with the proper error', ->
  assert.equal 'kaboom', assert.rejects fn
```

Note that if you want to be able to put more than one asynchronous test in a
single `it()`, you'll need to combine them somehow to make `mocha-as-promised`
happy, e.g.:

```coffee
{ runAsync } = require './some-library'
assert       = require 'assertive-as-promised'
Promise      = require 'bluebird'

it 'runs and fails asynchronously', ->
  Promise.all [
    assert.deepEqual 'got proper hash', { a: 42 }, runAsync('good')
    assert.rejects 'fails on bad', -> runAsync('bad')
  ]
```

(this may be bad style, depending on who you ask, but I find it useful if you
have `it()`s with a lot of setup overhead)

## Development

* `src/aap.coffee` is the main library; it compiles to `lib/aap.js`.
* `test/assertive_test.coffee` is a copy of the [assertive] library tests, slightly modified to run correctly in our test environment (see comments at the top)

[assertive]: https://github.com/groupon/assertive
[Mocha]: https://mochajs.org/


CHANGELOG.md
### 1.0.4

* Apply latest nlm generator - **[@markowsiak](https://github.com/markowsiak)** [#4](https://github.com/groupon/assertive-as-promised/pull/4)
  - [`8d9ebd1`](https://github.com/groupon/assertive-as-promised/commit/8d9ebd118d40993cbf8386b2a898575c6859fce6) **chore:** Apply latest nlm generator


### 1.0.3

* docs: announce deprecation due to assertive>2.1.0 - **[@dbushong](https://github.com/dbushong)** [#3](https://github.com/groupon/assertive-as-promised/pull/3)
  - [`e81494e`](https://github.com/groupon/assertive-as-promised/commit/e81494e2bfb7970ffc6f0d832df28490ddf97749) **docs:** announce deprecation due to assertive>2.1.0


### 1.0.2

* Apply latest nlm generator - **[@i-tier-bot](https://github.com/i-tier-bot)** [#2](https://github.com/groupon/assertive-as-promised/pull/2)
  - [`70fd106`](https://github.com/groupon/assertive-as-promised/commit/70fd1064c5e157f7c70d7e21701077c72eabfcf3) **chore:** Apply latest nlm generator
  - [`d514d90`](https://github.com/groupon/assertive-as-promised/commit/d514d900708084c09f622bd0a5edec29ae1686e3) **chore:** Use Groupon as author
  - [`c78d229`](https://github.com/groupon/assertive-as-promised/commit/c78d22990ca29864a9e3e4d244e319a15ad09f98) **test:** don't leak unhandled rejections during tests


## 1.0.1

* Switch to bluebird
* Upgrade requirements

## 0.10.0

* Initial revision using Q


lib/stable.js
// Generated by CoffeeScript 1.7.1

/*
Copyright (c) 2014, Groupon, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

Neither the name of GROUPON nor the names of its contributors may be
used to endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
var Q, assert, clear, emsg, fn, green, name, red, _fn,
  __slice = [].slice,
  __hasProp = {}.hasOwnProperty;

Q = require('q');

assert = require('assertive');

green = function(x) {
  return "\x1B[32m" + x + "\x1B[39m";
};

red = function(x) {
  return "\x1B[31m" + x + "\x1B[39m";
};

clear = "\x1b[39;49;00m";

emsg = function(message, explanation) {
  if (explanation != null) {
    message = "Assertion failed: " + explanation + "\n" + clear + message;
  }
  return message;
};

exports.rejects = function(expln, testee, name) {
  var _ref;
  if (name == null) {
    name = 'rejects';
  }
  if (testee != null) {
    assert.hasType("argument 1 of " + name + " must be a doc string", String, expln);
  } else {
    _ref = [expln, null], testee = _ref[0], expln = _ref[1];
  }
  if (!Q.isPromiseAlike(testee)) {
    if ('function' !== typeof testee) {
      throw new Error("" + name + " expects " + (green('a function or promise')) + " but got " + (red(testee)));
    }
    testee = Q["try"](testee);
  }
  if (name === 'rejects') {
    return testee.then(function() {
      throw new Error(emsg("Promise wasn't rejected as expected to", expln));
    }, function(err) {
      return err;
    });
  } else {
    return testee["catch"](function(err) {
      var _ref1;
      throw new Error(emsg("Promise was rejected despite resolves assertion:\n" + ((_ref1 = err != null ? err.message : void 0) != null ? _ref1 : err), expln));
    });
  }
};

exports.resolves = function(expln, testee) {
  return exports.rejects(expln, testee, 'resolves');
};

_fn = function(name, fn) {
  return exports[name] != null ? exports[name] : exports[name] = function() {
    var args, testee;
    args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    if (!args.length) {
      return fn();
    }
    testee = args.pop();
    if (Q.isPromiseAlike(testee)) {
      return testee.then(function(val) {
        return fn.apply(null, __slice.call(args).concat([val]));
      });
    } else {
      return fn.apply(null, __slice.call(args).concat([testee]));
    }
  };
};
for (name in assert) {
  if (!__hasProp.call(assert, name)) continue;
  fn = assert[name];
  _fn(name, fn);
}


lib/aap.js

/*
Copyright (c) 2014, Groupon, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

Neither the name of GROUPON nor the names of its contributors may be
used to endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
var Promise, aap, assert, clear, emsg, fn, fn1, green, isPromiseAlike, name, red,
  slice = [].slice,
  hasProp = {}.hasOwnProperty;

Promise = require('bluebird');

assert = require('assertive');

aap = {};

green = function(x) {
  return "\x1B[32m" + x + "\x1B[39m";
};

red = function(x) {
  return "\x1B[31m" + x + "\x1B[39m";
};

clear = "\x1b[39;49;00m";

emsg = function(message, explanation) {
  if (explanation != null) {
    message = "Assertion failed: " + explanation + "\n" + clear + message;
  }
  return message;
};

isPromiseAlike = function(p) {
  return p === Object(p) && 'function' === typeof p.then;
};

aap.rejects = function(expln, testee, name) {
  var ref;
  if (name == null) {
    name = 'rejects';
  }
  if (testee != null) {
    assert.hasType("argument 1 of " + name + " must be a doc string", String, expln);
  } else {
    ref = [expln, null], testee = ref[0], expln = ref[1];
  }
  if (!isPromiseAlike(testee)) {
    if ('function' !== typeof testee) {
      throw new Error(name + " expects " + (green('a function or promise')) + " but got " + (red(testee)));
    }
    testee = Promise["try"](testee);
  }
  if (name === 'rejects') {
    return testee.then(function() {
      throw new Error(emsg("Promise wasn't rejected as expected to", expln));
    }, function(err) {
      return err;
    });
  } else {
    return testee["catch"](function(err) {
      var ref1;
      throw new Error(emsg("Promise was rejected despite resolves assertion:\n" + ((ref1 = err != null ? err.message : void 0) != null ? ref1 : err), expln));
    });
  }
};

aap.resolves = function(expln, testee) {
  return aap.rejects(expln, testee, 'resolves');
};

fn1 = function(name, fn) {
  return aap[name] != null ? aap[name] : aap[name] = function() {
    var args, testee;
    args = 1 <= arguments.length ? slice.call(arguments, 0) : [];
    if (!args.length) {
      return fn();
    }
    testee = args.pop();
    if (isPromiseAlike(testee)) {
      return testee.then(function(val) {
        return fn.apply(null, slice.call(args).concat([val]));
      });
    } else {
      return fn.apply(null, slice.call(args).concat([testee]));
    }
  };
};
for (name in assert) {
  if (!hasProp.call(assert, name)) continue;
  fn = assert[name];
  fn1(name, fn);
}

if ((typeof module !== "undefined" && module !== null ? module.exports : void 0) != null) {
  module.exports = aap;
} else {
  global.assert = aap;
}
